lib/AFP/OsString.hs

```diff
@@ -0,0 +1,54 @@
+-- |
+-- Module      :  OsString
+-- Copyright   :  © 2021 Julian Ospald
+-- License     :  MIT
+--
+-- Maintainer  :  Julian Ospald <hasufell@posteo.de>
+-- Stability   :  experimental
+-- Portability :  portable
+--
+-- An implementation of platform specific 'OsString', which is:
```

Comment:
So this is the main deviation from the original spec. `OsString` is slightly more general than `AbstractFilePath`. The main argument is functions like:

```
splitSearchPath :: String -> [FilePath]
splitExtension :: FilePath -> (String, String)
```

They already return `String` here. Which means they are not necessarily *filepaths*. In addition, there may be unix or Win32 APIs that expect other input strings with the same encoding semantics as filepaths.

This follows the rust OsString model somewhat: https://doc.rust-lang.org/std/ffi/struct.OsString.html